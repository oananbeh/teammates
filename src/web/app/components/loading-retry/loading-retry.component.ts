import { Component, EventEmitter, Input, Output } from '@angular/core';

/**
 * Retry button for when loading fails.
 */
@Component({
  selector: 'tm-loading-retry',
  templateUrl: './loading-retry.component.html',
  styleUrls: ['./loading-retry.component.scss'],
})
export class LoadingRetryComponent {

  @Input() message: string = '';

  @Input() shouldShowRetry: boolean = false;

  @Input() retryButtonDisabled: boolean = false;

  @Output() retryEvent: EventEmitter<any> = new EventEmitter<any>();

}
