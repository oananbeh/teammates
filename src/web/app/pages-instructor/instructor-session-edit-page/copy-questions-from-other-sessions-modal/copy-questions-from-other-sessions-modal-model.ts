import { FeedbackQuestion } from '../../../../types/api-output';
import { SortBy, SortOrder } from '../../../../types/sort-properties';

/**
 * The model for a feedback session tab.
 */
export interface FeedbackSessionTabModel {
  courseId: string;
  feedbackSessionName: string;
  createdAtTimestamp: number;

  questionsTableRowModels: QuestionToCopyCandidate[];
  questionsTableRowModelsSortBy: SortBy;
  questionsTableRowModelsSortOrder: SortOrder;

  hasQuestionsLoaded: boolean;
  isTabExpanded: boolean;
  hasLoadingFailed: boolean;
}

/**
 * The model for a question to copy.
 */
export interface QuestionToCopyCandidate {
  question: FeedbackQuestion;
  isSelected: boolean;
}
