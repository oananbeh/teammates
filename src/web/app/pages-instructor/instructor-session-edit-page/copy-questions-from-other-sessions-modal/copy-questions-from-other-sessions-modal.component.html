<div class="modal-header">
  <h4 class="modal-title">Copy Questions</h4>
  <button type="button" class="btn-close" aria-label="Close" (click)="activeModal.dismiss()"></button>
</div>
<div id="copy-question-modal" class="modal-body">
  <div class="col-12 mb-3 text-md-end">
    <strong class="d-inline"> Sort Sessions By: </strong>
    <div class="btn-group" data-toggle="buttons">
      <button id="sort-course-id" class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.COURSE_ID)" (click)="sortFeedbackSessionTabs(SortBy.COURSE_ID)">Course ID</button>
      <button id="sort-session-name" class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.SESSION_NAME)" (click)="sortFeedbackSessionTabs(SortBy.SESSION_NAME)">Session Name</button>
      <button id="sort-creation-date" class="btn btn-light" [disabled]="isSelectedForSorting(SortBy.SESSION_CREATION_DATE)" (click)="sortFeedbackSessionTabs(SortBy.SESSION_CREATION_DATE)">Creation Date</button>
    </div>
  </div>
  <div class="feedback-session-tabs" *ngFor="let feedbackSessionTabModel of feedbackSessionTabModels">
    <div class="card">
      <div class="card-header cursor-pointer bg-primary text-white" (click)="toggleCard(feedbackSessionTabModel)">
        <strong class="text-break">[{{feedbackSessionTabModel.courseId}}] Session: {{feedbackSessionTabModel.feedbackSessionName}}</strong>
        <div class="card-header-btn-toolbar">
          <tm-panel-chevron [isExpanded]="feedbackSessionTabModel.isTabExpanded"></tm-panel-chevron>
        </div>
      </div>

      <div *ngIf="feedbackSessionTabModel.isTabExpanded">
        <tm-loading-retry [shouldShowRetry]="feedbackSessionTabModel.hasLoadingFailed" [message]="'Failed to load questions'" (retryEvent)="loadQuestions(feedbackSessionTabModel)">
          <div *tmIsLoading="!feedbackSessionTabModel.hasQuestionsLoaded">
            <div *ngIf="!feedbackSessionTabModel.hasLoadingFailed" class="card-body">
              <ng-container *ngIf="feedbackSessionTabModel.hasQuestionsLoaded && feedbackSessionTabModel.questionsTableRowModels.length > 0, else noQuestionsTemplate">
                <table id="copy-question-table" class="table table-bordered table-responsive-lg">
                  <thead class="bg-primary text-white">
                    <tr>
                      <th>#</th>
                      <th class="sortable-header" (click)="sortQuestionsToCopyForFeedbackSession(feedbackSessionTabModel, SortBy.QUESTION_TYPE)" [attr.aria-sort]="getAriaSort(feedbackSessionTabModel, SortBy.QUESTION_TYPE)">
                        <button>
                          Question Type
                          <span class="fa-stack" aria-hidden="true">
                            <i class="fas fa-sort"></i>
                            <i *ngIf="isSortQuestionsBy(feedbackSessionTabModel, SortBy.QUESTION_TYPE, SortOrder.DESC)" class="fas fa-sort-down"></i>
                            <i *ngIf="isSortQuestionsBy(feedbackSessionTabModel, SortBy.QUESTION_TYPE, SortOrder.ASC)" class="fas fa-sort-up"></i>
                          </span>
                        </button>
                      </th>
                      <th class="sortable-header" (click)="sortQuestionsToCopyForFeedbackSession(feedbackSessionTabModel, SortBy.QUESTION_TEXT)" [attr.aria-sort]="getAriaSort(feedbackSessionTabModel, SortBy.QUESTION_TEXT)">
                        <button>
                          Question Text
                          <span class="fa-stack" aria-hidden="true">
                            <i class="fas fa-sort"></i>
                            <i *ngIf="isSortQuestionsBy(feedbackSessionTabModel, SortBy.QUESTION_TEXT, SortOrder.DESC)" class="fas fa-sort-down"></i>
                            <i *ngIf="isSortQuestionsBy(feedbackSessionTabModel, SortBy.QUESTION_TEXT, SortOrder.ASC)" class="fas fa-sort-up"></i>
                          </span>
                        </button>
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr *ngFor="let questionCandidate of feedbackSessionTabModel.questionsTableRowModels">
                      <td><input type="checkbox" [(ngModel)]="questionCandidate.isSelected"></td>
                      <td>{{ questionCandidate.question.questionType | questionTypeName }}</td>
                      <td>{{ questionCandidate.question.questionBrief }}</td>
                    </tr>
                  </tbody>
                </table>
              </ng-container>
              <button (click)="loadQuestions(feedbackSessionTabModel)" class="btn btn-primary float-end mt-2 mb-3">
                <i class="fas fa-redo" aria-hidden="true"></i>
                &nbsp;Reload
              </button>
            </div>
          </div>
        </tm-loading-retry>
      </div>

      <ng-template #noQuestionsTemplate>
        <h5>There are no questions in this feedback session available for copying.</h5>
      </ng-template>

    </div>
  </div>
</div>
<div class="modal-footer">
  <button id="btn-confirm-copy-question" type="button" class="btn btn-primary" (click)="activeModal.close(getSelectedQuestions())"
          [disabled]="!hasAnyQuestionsToCopySelected">Copy</button>
  <button type="button" class="btn btn-light" (click)="activeModal.dismiss()">Cancel</button>
</div>
