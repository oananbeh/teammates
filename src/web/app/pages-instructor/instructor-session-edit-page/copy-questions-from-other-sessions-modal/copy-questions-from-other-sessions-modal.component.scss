.sortable-header {
  cursor: pointer;

  i {
    margin-left: 0.3em;
  }

  .fa-stack {
    width: 20%;
  }

  button {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    text-align: left;
    width: 100%;
    font: inherit;
    cursor: inherit;
  }
}

#copy-question-modal {
  table-layout: auto;
  margin-bottom: 10px;
  min-width: 70vw;
}

.modal-footer {
  justify-content: right;
  background-color: white;
}
