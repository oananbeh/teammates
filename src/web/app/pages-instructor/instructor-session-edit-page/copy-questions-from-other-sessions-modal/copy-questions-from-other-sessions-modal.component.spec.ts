import { CommonModule } from '@angular/common';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';
import { FeedbackSessionTabModel, QuestionToCopyCandidate } from './copy-questions-from-other-sessions-modal-model';
import { CopyQuestionsFromOtherSessionsModalComponent } from './copy-questions-from-other-sessions-modal.component';
import { FeedbackQuestionsService } from '../../../../services/feedback-questions.service';
import {
  FeedbackMcqQuestionDetails,
  FeedbackParticipantType,
  FeedbackQuestion,
  FeedbackQuestions,
  FeedbackQuestionType,
  FeedbackRankRecipientsQuestionDetails,
  FeedbackTextQuestionDetails,
  FeedbackVisibilityType,
  NumberOfEntitiesToGiveFeedbackToSetting,
} from '../../../../types/api-output';
import { SortBy, SortOrder } from '../../../../types/sort-properties';
import { LoadingRetryModule } from '../../../components/loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../../../components/loading-spinner/loading-spinner.module';
import { PanelChevronModule } from '../../../components/panel-chevron/panel-chevron.module';
import { TeammatesCommonModule } from '../../../components/teammates-common/teammates-common.module';

describe('CopyQuestionsFromOtherSessionsModalComponent', () => {

  const testFeedbackQuestion1: FeedbackQuestion = {
    feedbackQuestionId: 'feedback-question-1',
    questionNumber: 1,
    questionBrief: 'question brief',
    questionDescription: 'description',
    questionType: FeedbackQuestionType.TEXT,
    questionDetails: {
      questionType: FeedbackQuestionType.TEXT,
      questionText: 'question text',
    } as FeedbackTextQuestionDetails,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.INSTRUCTORS,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    customNumberOfEntitiesToGiveFeedbackTo: 5,
    showResponsesTo: [FeedbackVisibilityType.GIVER_TEAM_MEMBERS, FeedbackVisibilityType.INSTRUCTORS],
    showGiverNameTo: [],
    showRecipientNameTo: [],
  };

  const testFeedbackQuestion2: FeedbackQuestion = {
    feedbackQuestionId: 'feedback-question-2',
    questionNumber: 2,
    questionBrief: 'question brief',
    questionDescription: 'description',
    questionType: FeedbackQuestionType.MCQ,
    questionDetails: {
      questionType: FeedbackQuestionType.MCQ,
      questionText: 'question text',
      mcqChoices: ['choice 1', 'choice 2', 'choice 3'],
    } as FeedbackMcqQuestionDetails,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.INSTRUCTORS,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    customNumberOfEntitiesToGiveFeedbackTo: 5,
    showResponsesTo: [FeedbackVisibilityType.GIVER_TEAM_MEMBERS, FeedbackVisibilityType.INSTRUCTORS],
    showGiverNameTo: [],
    showRecipientNameTo: [],
  };

  const testFeedbackQuestion3: FeedbackQuestion = {
    feedbackQuestionId: 'feedback-question-3',
    questionNumber: 3,
    questionBrief: 'question brief',
    questionDescription: 'description',
    questionType: FeedbackQuestionType.RANK_RECIPIENTS,
    questionDetails: {
      questionType: FeedbackQuestionType.RANK_RECIPIENTS,
      questionText: 'question text',
      minOptionsToBeRanked: 5,
      maxOptionsToBeRanked: 5,
      areDuplicatesAllowed: true,
    } as FeedbackRankRecipientsQuestionDetails,
    giverType: FeedbackParticipantType.STUDENTS,
    recipientType: FeedbackParticipantType.INSTRUCTORS,
    numberOfEntitiesToGiveFeedbackToSetting: NumberOfEntitiesToGiveFeedbackToSetting.UNLIMITED,
    customNumberOfEntitiesToGiveFeedbackTo: 5,
    showResponsesTo: [FeedbackVisibilityType.GIVER_TEAM_MEMBERS, FeedbackVisibilityType.INSTRUCTORS],
    showGiverNameTo: [],
    showRecipientNameTo: [],
  };

  const testQuestionToCopyCandidate1: QuestionToCopyCandidate = {
    question: testFeedbackQuestion1,
    isSelected: false,
  };

  const testQuestionToCopyCandidate2: QuestionToCopyCandidate = {
    question: testFeedbackQuestion2,
    isSelected: false,
  };

  const testQuestionToCopyCandidate3: QuestionToCopyCandidate = {
    question: testFeedbackQuestion3,
    isSelected: false,
  };

  const testFeedbackSessionTabModel1: FeedbackSessionTabModel = {
    courseId: 'testId1',
    feedbackSessionName: 'Test Session 1',
    createdAtTimestamp: 1644458400, // Thursday, 10 February 2022 10:00:00 GMT+08:00
    questionsTableRowModels: [],
    questionsTableRowModelsSortBy: SortBy.NONE,
    questionsTableRowModelsSortOrder: SortOrder.ASC,

    hasQuestionsLoaded: false,
    isTabExpanded: false,
    hasLoadingFailed: false,
  };

  const testFeedbackSessionTabModel2: FeedbackSessionTabModel = {
    courseId: 'testId2',
    feedbackSessionName: 'Test Session 2',
    createdAtTimestamp: 1645063200, // Thursday, 17 February 2022 10:00:00 GMT+08:00
    questionsTableRowModels: [],
    questionsTableRowModelsSortBy: SortBy.NONE,
    questionsTableRowModelsSortOrder: SortOrder.ASC,

    hasQuestionsLoaded: false,
    isTabExpanded: false,
    hasLoadingFailed: false,
  };

  let component: CopyQuestionsFromOtherSessionsModalComponent;
  let fixture: ComponentFixture<CopyQuestionsFromOtherSessionsModalComponent>;
  let feedbackQuestionsService: FeedbackQuestionsService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [CopyQuestionsFromOtherSessionsModalComponent],
      imports: [
        CommonModule,
        FormsModule,
        TeammatesCommonModule,
        PanelChevronModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
        HttpClientTestingModule,
      ],
      providers: [
        NgbActiveModal,
        FeedbackQuestionsService,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CopyQuestionsFromOtherSessionsModalComponent);
    feedbackQuestionsService = TestBed.inject(FeedbackQuestionsService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with feedback sessions', () => {
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1, testFeedbackSessionTabModel2];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when feedback questions are loading', () => {
    testFeedbackSessionTabModel1.isTabExpanded = true;
    testFeedbackSessionTabModel1.hasQuestionsLoaded = false;
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when feedback questions are loaded', () => {
    testFeedbackSessionTabModel1.isTabExpanded = true;
    testFeedbackSessionTabModel1.hasQuestionsLoaded = true;
    testFeedbackSessionTabModel1.questionsTableRowModels = [testQuestionToCopyCandidate1, testQuestionToCopyCandidate2];
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when feedback questions failed to load', () => {
    testFeedbackSessionTabModel1.isTabExpanded = true;
    testFeedbackSessionTabModel1.hasLoadingFailed = true;
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should load feedback questions', () => {
    const feedbackQuestions: FeedbackQuestions = {
      questions: [testFeedbackQuestion1, testFeedbackQuestion2],
    };
    jest.spyOn(feedbackQuestionsService, 'getFeedbackQuestions').mockReturnValue(of(feedbackQuestions));
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1];

    component.loadQuestions(testFeedbackSessionTabModel1);

    expect(component.feedbackSessionTabModels[0].hasQuestionsLoaded).toBeTruthy();
    expect(component.feedbackSessionTabModels[0].hasLoadingFailed).toBeFalsy();
    expect(component.feedbackSessionTabModels[0].questionsTableRowModels.length).toBe(2);
    expect(component.feedbackSessionTabModels[0].questionsTableRowModels[0].question.feedbackQuestionId)
      .toBe(testFeedbackQuestion1.feedbackQuestionId);
    expect(component.feedbackSessionTabModels[0].questionsTableRowModels[1].question.feedbackQuestionId)
      .toBe(testFeedbackQuestion2.feedbackQuestionId);
  });

  it('should not allow copying when no questions are selected', () => {
    testFeedbackSessionTabModel1.questionsTableRowModels = [testQuestionToCopyCandidate1, testQuestionToCopyCandidate2];
    testFeedbackSessionTabModel2.questionsTableRowModels = [testQuestionToCopyCandidate3];
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1, testFeedbackSessionTabModel2];
    fixture.detectChanges();

    const questions: FeedbackQuestion[] = component.getSelectedQuestions();
    const button: any = fixture.debugElement.nativeElement.querySelector('#btn-confirm-copy-question');

    expect(component.hasAnyQuestionsToCopySelected).toBeFalsy();
    expect(questions.length).toBe(0);
    expect(button.disabled).toBeTruthy();
  });

  it('should copy selected questions', () => {
    testQuestionToCopyCandidate1.isSelected = true;
    testQuestionToCopyCandidate2.isSelected = true;
    testQuestionToCopyCandidate3.isSelected = true;
    testFeedbackSessionTabModel1.questionsTableRowModels = [testQuestionToCopyCandidate1, testQuestionToCopyCandidate2];
    testFeedbackSessionTabModel2.questionsTableRowModels = [testQuestionToCopyCandidate3];
    component.feedbackSessionTabModels = [testFeedbackSessionTabModel1, testFeedbackSessionTabModel2];
    fixture.detectChanges();

    jest.spyOn(component.activeModal, 'close').mockImplementation((questions: FeedbackQuestion[]) => {
      expect(questions.length).toBe(3);
      expect(questions[0].feedbackQuestionId).toBe(testFeedbackQuestion1.feedbackQuestionId);
      expect(questions[1].feedbackQuestionId).toBe(testFeedbackQuestion2.feedbackQuestionId);
      expect(questions[2].feedbackQuestionId).toBe(testFeedbackQuestion3.feedbackQuestionId);
    });

    const button: any = fixture.debugElement.nativeElement.querySelector('#btn-confirm-copy-question');

    expect(component.hasAnyQuestionsToCopySelected).toBeTruthy();
    expect(button.disabled).toBeFalsy();

    button.click();
    expect(component.activeModal.close).toHaveBeenCalledTimes(1);
  });
});
