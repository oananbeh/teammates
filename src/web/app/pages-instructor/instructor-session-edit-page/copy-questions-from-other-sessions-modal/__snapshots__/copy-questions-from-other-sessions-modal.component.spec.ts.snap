// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyQuestionsFromOtherSessionsModalComponent should snap when feedback questions are loaded 1`] = `
<tm-copy-questions-from-other-sessions-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionTabModels={[Function Array]}
  feedbackSessionTabModelsSortBy={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Questions
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-question-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Sessions By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          disabled=""
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-session-name"
        >
          Session Name
        </button>
        <button
          class="btn btn-light"
          id="sort-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div
      class="feedback-session-tabs"
    >
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [testId1] Session: Test Session 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div>
          <tm-loading-retry>
            <div>
              <div
                class="card-body"
              >
                <table
                  class="table table-bordered table-responsive-lg"
                  id="copy-question-table"
                >
                  <thead
                    class="bg-primary text-white"
                  >
                    <tr>
                      <th>
                        #
                      </th>
                      <th
                        aria-sort="none"
                        class="sortable-header"
                      >
                        <button>
                           Question Type 
                          <span
                            aria-hidden="true"
                            class="fa-stack"
                          >
                            <i
                              class="fas fa-sort"
                            />
                          </span>
                        </button>
                      </th>
                      <th
                        aria-sort="none"
                        class="sortable-header"
                      >
                        <button>
                           Question Text 
                          <span
                            aria-hidden="true"
                            class="fa-stack"
                          >
                            <i
                              class="fas fa-sort"
                            />
                          </span>
                        </button>
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                        <input
                          class="ng-untouched ng-pristine ng-valid"
                          type="checkbox"
                        />
                      </td>
                      <td>
                        Essay question
                      </td>
                      <td>
                        question brief
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <input
                          class="ng-untouched ng-pristine ng-valid"
                          type="checkbox"
                        />
                      </td>
                      <td>
                        Multiple-choice (single answer) question
                      </td>
                      <td>
                        question brief
                      </td>
                    </tr>
                  </tbody>
                </table>
                <button
                  class="btn btn-primary float-end mt-2 mb-3"
                >
                  <i
                    aria-hidden="true"
                    class="fas fa-redo"
                  />
                   Â Reload 
                </button>
              </div>
            </div>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-question"
      type="button"
    >
      Copy
    </button>
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-questions-from-other-sessions-modal>
`;

exports[`CopyQuestionsFromOtherSessionsModalComponent should snap when feedback questions are loading 1`] = `
<tm-copy-questions-from-other-sessions-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionTabModels={[Function Array]}
  feedbackSessionTabModelsSortBy={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Questions
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-question-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Sessions By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          disabled=""
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-session-name"
        >
          Session Name
        </button>
        <button
          class="btn btn-light"
          id="sort-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div
      class="feedback-session-tabs"
    >
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [testId1] Session: Test Session 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div>
          <tm-loading-retry>
            <tm-loading-spinner>
              <div
                class="loading-container"
              >
                <div
                  class="row spinner-border text-primary"
                  role="status"
                />
                <div
                  class="row text-secondary"
                >
                  Loading...
                </div>
              </div>
            </tm-loading-spinner>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-question"
      type="button"
    >
      Copy
    </button>
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-questions-from-other-sessions-modal>
`;

exports[`CopyQuestionsFromOtherSessionsModalComponent should snap when feedback questions failed to load 1`] = `
<tm-copy-questions-from-other-sessions-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionTabModels={[Function Array]}
  feedbackSessionTabModelsSortBy={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Questions
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-question-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Sessions By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          disabled=""
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-session-name"
        >
          Session Name
        </button>
        <button
          class="btn btn-light"
          id="sort-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div
      class="feedback-session-tabs"
    >
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [testId1] Session: Test Session 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div>
          <tm-loading-retry>
            <div
              class="text-center"
            >
              <p
                class="text-muted m-1"
              >
                Failed to load questions
              </p>
              <button
                class="btn btn-primary"
              >
                <i
                  aria-hidden="true"
                  class="fas fa-redo"
                />
                Â Retry
              </button>
            </div>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-question"
      type="button"
    >
      Copy
    </button>
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-questions-from-other-sessions-modal>
`;

exports[`CopyQuestionsFromOtherSessionsModalComponent should snap with default fields 1`] = `
<tm-copy-questions-from-other-sessions-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionTabModels={[Function Array]}
  feedbackSessionTabModelsSortBy={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Questions
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-question-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Sessions By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          disabled=""
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-session-name"
        >
          Session Name
        </button>
        <button
          class="btn btn-light"
          id="sort-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-question"
      type="button"
    >
      Copy
    </button>
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-questions-from-other-sessions-modal>
`;

exports[`CopyQuestionsFromOtherSessionsModalComponent should snap with feedback sessions 1`] = `
<tm-copy-questions-from-other-sessions-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  feedbackQuestionsService={[Function FeedbackQuestionsService]}
  feedbackSessionTabModels={[Function Array]}
  feedbackSessionTabModelsSortBy={[Function Number]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Questions
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-question-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Sessions By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          disabled=""
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-session-name"
        >
          Session Name
        </button>
        <button
          class="btn btn-light"
          id="sort-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div
      class="feedback-session-tabs"
    >
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [testId1] Session: Test Session 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="false"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-down"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
      </div>
    </div>
    <div
      class="feedback-session-tabs"
    >
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [testId2] Session: Test Session 2
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="false"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-down"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-question"
      type="button"
    >
      Copy
    </button>
    <button
      class="btn btn-light"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-questions-from-other-sessions-modal>
`;
